Add revocation certificate checks

pepone opened the pull request:
This PR adds certificate revocation checks for IceSSL C++

bentoi commented on cpp/test/IceSSL/configuration/AllTests.cpp:
```diff
@@ -4019,6 +4019,131 @@ allTests(Test::TestHelper* helper, const string& /*testDir*/, bool p12)
+        fact->destroyServer(server);
+        comm->destroy();
+
+        // Test with s_rsa_cai3 only the intermeidate CA cert is revoked
```
intermediate

bentoi reviewed (approved):
It's not clear what process provides the CRL on port 20001. There's no way to test CRL with secure transport? 

Also, what about blocking blocking network fetch for the CRL? I guess this still needs to be looked into?

pepone commented:
> It's not clear what process provides the CRL on port 20001. There's no way to test CRL with secure transport?

Pushed the scripts changes, see `scripts/tests/IceSSL/configuration.py`, I think using HTTP is the norm for CRL, the CRL is cryptographically signed. See https://security.stackexchange.com/questions/242953/crl-over-https-is-it-really-a-bad-practice/242959#242959

> Also, what about blocking blocking network fetch for the CRL? I guess this still needs to be looked into?

Yes this is not done yet

pepone commented on cpp/src/IceSSL/OpenSSLEngine.cpp:
```diff
@@ -816,14 +816,72 @@ OpenSSL::SSLEngine::initialize()
                 }
             }
 
+            string crlFile = properties->getProperty(propPrefix + "CRLFile");
```
Fixed

pepone commented on cpp/src/IceSSL/OpenSSLEngine.cpp:
```diff
@@ -816,14 +816,72 @@ OpenSSL::SSLEngine::initialize()
+                }
+
+                int revocationCheck = getRevocationCheck();
+                if(revocationCheck > 0)
```
yes fixed

externl reviewed (approved):
Looks good to me!

pepone commented:
Added OCSP revocation tests supported in macOS and Windows, I also updated SChannel `trustStatusToTrustError` implementation to report more critical errors first as it can only report one error.

bentoi commented on cpp/src/IceSSL/SChannelTransceiverI.cpp:
```diff
@@ -765,8 +762,22 @@ SChannel::TransceiverI::initialize(IceInternal::Buffer& readBuffer, IceInternal:
         PCCERT_CHAIN_CONTEXT certChain;
-        if(!CertGetCertificateChain(_engine->chainEngine(), cert, 0, cert->hCertStore, &chainP,
-                                    CERT_CHAIN_REVOCATION_CHECK_CACHE_ONLY, 0, &certChain))
+        DWORD dwFlags = 0;
```
Are we changing the default behavior here? Or does the cache only flag had no effect if the revocation flags weren't set?

pepone replied:
Yes I think this is only relevant if revocation checks are enabled, this is the same that .NET runtime does

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs#L124-L139

bentoi commented on cpp/src/IceSSL/SecureTransportTransceiverI.cpp:
```diff
@@ -155,37 +155,65 @@ checkTrustResult(SecTrustRef trust,
             {
                 throw SecurityException(__FILE__, __LINE__, "IceSSL: handshake failure:\n" + sslErrorToString(err));
             }
+            SecTrustSetAnchorCertificatesOnly(trust, true);
```
It looks like we didn't set this before, will this have an effect on binary compatibility?

pepone replied:
This API tells SecureTransport to only trust the certificate authorities set above, I think this is the intended behaviour of `IceSSL.CAs, I think it is safe to update this.

bentoi commented on cpp/src/IceSSL/SecureTransportTransceiverI.cpp:
```diff
@@ -155,37 +155,65 @@ checkTrustResult(SecTrustRef trust,
-        //
-        // Disable network fetch, we don't want this to block.
-        //
+        // Do not allow to fetch missing intermediate certificates from the network.
```
Is this consistent with other SSL implementations? All of them forbids fetching intermediate certificates?

pepone replied:
OpenSSL doesn't fetch anything, for SChannel seems that the equivalent is `CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL` but we currently don't set this flag. Note that I change the comment because this doesn't affect OCSP requests

bentoi commented:
OK, sounds good.
